### PR TITLE
Relax shapes for Keras _on_batch functions.

### LINKED ISSUE #34907: [TF 2.1rc0] The `{predict,train,test}_on_batch` trace functions with fixed batch size
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian Stretch
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.1.0rc0
- Python version: 3.5.3

**Describe the current behavior**

When using the updated `predict_on_batch` method wrapped now in `tf.function`, the traced calls have a fixed batch size. Therefore, if batch size changes regularly, the function is retraced all the time. For example, the following

```python
import numpy as np
import tensorflow as tf

model = tf.keras.Sequential([
    tf.keras.layers.Dense(1, input_shape=[1])
])
model.compile(optimizer=tf.optimizers.Adam(),
              loss=tf.losses.MeanSquaredError())

for i in range(1, 300):
    model.predict_on_batch(np.ones([i, 1]))
```

gives

```
WARNING:tensorflow:5 out of the last 5 calls to <tensorflow.python.keras.engine.sequential.Sequential object at 0x7f2185a37630> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/tutorials/customization/performance#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.
WARNING:tensorflow:6 out of the last 6 calls to <tensorflow.python.keras.engine.sequential.Sequential object at 0x7f2185a37630> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/tutorials/customization/performance#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.
...
```

The same issue happend with `train_on_batch` and `test_on_batch`.

**Describe the expected behavior**

The `{predict,train,test}_on_batch` methods are traced with undefined batch size.

Alternatively, the behaviour could be configurable. Each of `{predict,train,test}_on_batch` could take `undefined_batch_size=True/False` argument; personally I vote for a default value of `undefined_batch_size=True`.

**And I just realized:** if `*_on_batch` methods are used with sequences (for example sentences in NLP), then the retracing would happen on every call. So instead, it would be better to allow passing `experimental_relax_shapes` to the `*_on_batch` or even to `.compile` call.

**Code to reproduce the issue**
Available above.

**Other info / logs**

### PR DESCRIPTION
**Pull request by foxik**
EDIT(robieta): Moved below.

### CODE REVIEW

File: tensorflow/python/keras/engine/training_v2_utils.py [hunk 1]
```diff
@@ -119,7 +119,8 @@ def _make_on_batch_function(model, mode):
     func = model
 
   if not model.run_eagerly:
-    func = def_function.function(func)
+    # Pass the experimental_relax_shapes to gracefully handle dynamic batch size.
```
Comment (robieta) [inline tensorflow/python/keras/engine/training_v2_utils.py]: I would reword this to:
```
# Pass `experimental_relax_shapes` to avoid retracing for dynamic batch size,
# variable length sequences, etc.
```

**Review by robieta — COMMENTED** [review]
Minor tweak on the comment, but otherwise looks good.

File: tensorflow/python/keras/engine/training_v2_utils.py (hunk 1 above)
Comment (foxik) [reply tensorflow/python/keras/engine/training_v2_utils.py]: Updated as suggested.

**Comment by foxik** [comment]
Updated as suggested.

**Review by robieta — DISMISSED** [review]

**Comment by robieta** [comment]
Original description:

The current `{train,test,predict}_on_batch` function use a regular `tf.function` when not in eager mode, which causes a retrace for every new batch size. Similarly, if sequences are passed on input, every different sequence size causes a retrace.

Passing experimental_relax_shapes=True allow gracefully handle these cases.

Fixes #34907.

**Comment by robieta** [comment]
I'm currently running into some infra issue getting this in. I will update when it's resolved.

**Comment by robieta** [comment]
@foxik can you rewrite history to strip the commit message down to something simple? There seems to be a special character in there somewhere which is messing with our infra. Sorry for the inconvenience.

**Comment by foxik** [comment]
@robieta Done, I removed all weird characters (there were an asterisk, braces and backticks), and left only underscores and an equal sign, which seem fairly benign.

BTW, sorry for the delay, but I am in UTC+1 timezone :-)
